File: src/Services/Implementations/CartService.cs
using ECommerceApi.DTOs.Cart;
using ECommerceApi.Exceptions;
using ECommerceApi.Models;
using ECommerceApi.Repositories.Interfaces;
using ECommerceApi.Services.Interfaces;

namespace ECommerceApi.Services.Implementations;

public class CartService : ICartService
{
    private readonly ICartRepository _cartRepository;
    private readonly IProductRepository _productRepository;

    public CartService(ICartRepository cartRepository, IProductRepository productRepository)
    {
        _cartRepository = cartRepository;
        _productRepository = productRepository;
    }

    public async Task<CartResponse> GetCartAsync(Guid userId)
    {
        var cart = await _cartRepository.FindOrCreateByUserAsync(userId);
        return await MapAsync(cart);
    }

    public async Task<CartResponse> AddItemAsync(Guid userId, Guid productId, int quantity)
    {
        if (quantity <= 0)
        {
            throw new ValidationException("Quantity must be a positive integer");
        }

        var cart = await _cartRepository.FindOrCreateByUserAsync(userId);

        var product = await _productRepository.FindByIdAsync(productId);
        if (product is null)
        {
            throw new NotFoundAppException("Product");
        }

        var existing = await _cartRepository.FindItemAsync(cart.Id, productId);
        var desiredQuantity = (existing?.Quantity ?? 0) + quantity;

        if (desiredQuantity > product.StockQuantity)
        {
            throw new ValidationException($"Only {product.StockQuantity} unit(s) of \"{product.Name}\" are available");
        }

        var updated = existing is not null
            ? await _cartRepository.UpdateItemQuantityAsync(cart.Id, existing.Id, desiredQuantity)
            : await _cartRepository.AddItemAsync(cart.Id, productId, quantity);

        return await MapAsync(updated);
    }

    /*
      Updates the quantity of a product already in the user's cart.
      Addressed by <paramref name="productId"/>, not the CartItem's own id
    */
    public async Task<CartResponse> UpdateItemQuantityAsync(Guid userId, Guid productId, int quantity)
    {
        if (quantity < 0)
        {
            throw new ValidationException("Quantity must be zero or a positive integer");
        }

        var cart = await _cartRepository.FindOrCreateByUserAsync(userId);
        var item = cart.Items.FirstOrDefault(i => i.ProductId == productId);
        if (item is null)
        {
            throw new NotFoundAppException("Cart item");
        }

        if (quantity == 0)
        {
            var cleared = await _cartRepository.RemoveItemAsync(cart.Id, item.Id);
            return await MapAsync(cleared);
        }

        var product = await _productRepository.FindByIdAsync(productId);
        if (product is null)
        {
            throw new NotFoundAppException("Product");
        }
        if (quantity > product.StockQuantity)
        {
            throw new ValidationException($"Only {product.StockQuantity} unit(s) of \"{product.Name}\" are available");
        }

        var updated = await _cartRepository.UpdateItemQuantityAsync(cart.Id, item.Id, quantity);
        return await MapAsync(updated);
    }

    public async Task<CartResponse> RemoveItemAsync(Guid userId, Guid productId)
    {
        var cart = await _cartRepository.FindOrCreateByUserAsync(userId);
        var item = cart.Items.FirstOrDefault(i => i.ProductId == productId);
        if (item is null)
        {
            throw new NotFoundAppException("Cart item");
        }

        var updated = await _cartRepository.RemoveItemAsync(cart.Id, item.Id);
        return await MapAsync(updated);
    }

    private async Task<CartResponse> MapAsync(Cart cart)
    {
        var items = new List<CartItemResponse>();
        decimal subtotal = 0m;

        foreach (var item in cart.Items)
        {
            var product = item.Product ?? await _productRepository.FindByIdIncludingDeletedAsync(item.ProductId);
            var unitPrice = product?.Price ?? 0m;
            subtotal += unitPrice * item.Quantity;

            items.Add(new CartItemResponse
            {
                Id = item.Id,
                ProductId = item.ProductId,
                ProductName = product?.Name ?? "Unknown product",
                UnitPrice = unitPrice,
                Quantity = item.Quantity
            });
        }

        return new CartResponse
        {
            Id = cart.Id,
            UserId = cart.UserId,
            Items = items,
            Subtotal = subtotal
        };
    }
}

File: legacy-nodejs/src/services/cart.service.ts
import { ICartRepository, IProductRepository } from '../repositories/interfaces';
import { CartRecord } from '../types/domain';
import { NotFoundError, ValidationError } from '../errors/AppError';

export class CartService {
  constructor(
    private readonly cartRepository: ICartRepository,
    private readonly productRepository: IProductRepository
  ) {}

  async getCart(userId: string): Promise<CartRecord> {
    return this.cartRepository.findOrCreateByUser(userId);
  }

  async addItem(userId: string, productId: string, quantity: number): Promise<CartRecord> {
    if (!Number.isInteger(quantity) || quantity <= 0) {
      throw new ValidationError('Quantity must be a positive integer');
    }

    const cart = await this.cartRepository.findOrCreateByUser(userId);

    const product = await this.productRepository.findById(productId);
    if (!product || !product.isActive) {
      throw new NotFoundError('Product');
    }

    const existing = await this.cartRepository.findItem(cart.id, productId);
    const desiredQuantity = (existing?.quantity ?? 0) + quantity;

    if (desiredQuantity > product.stockQuantity) {
      throw new ValidationError(
        `Only ${product.stockQuantity} unit(s) of "${product.name}" are available`
      );
    }

    if (existing) {
      return this.cartRepository.updateItemQuantity(cart.id, existing.id, desiredQuantity);
    }
    return this.cartRepository.addItem(cart.id, productId, quantity);
  }

  async updateItemQuantity(userId: string, itemId: string, quantity: number): Promise<CartRecord> {
    if (!Number.isInteger(quantity) || quantity < 0) {
      throw new ValidationError('Quantity must be zero or a positive integer');
    }

    const cart = await this.cartRepository.findOrCreateByUser(userId);
    const item = cart.items.find((i) => i.id === itemId);
    if (!item) {
      throw new NotFoundError('Cart item');
    }

    if (quantity === 0) {
      return this.cartRepository.removeItem(cart.id, itemId);
    }

    const product = await this.productRepository.findById(item.productId);
    if (!product || !product.isActive) {
      throw new NotFoundError('Product');
    }
    if (quantity > product.stockQuantity) {
      throw new ValidationError(`Only ${product.stockQuantity} unit(s) of "${product.name}" are available`);
    }

    return this.cartRepository.updateItemQuantity(cart.id, itemId, quantity);
  }

  async removeItem(userId: string, itemId: string): Promise<CartRecord> {
    const cart = await this.cartRepository.findOrCreateByUser(userId);
    const item = cart.items.find((i) => i.id === itemId);
    if (!item) {
      throw new NotFoundError('Cart item');
    }
    return this.cartRepository.removeItem(cart.id, itemId);
  }
}
